scheduler/scheduler.go

```diff
@@ -75,89 +66,24 @@ func (s *Scheduler) ReserveResources(hostID string, force bool, resourceRequests
+	reserveActions := ProvideReserveAction()
 	data := map[string]interface{}{}
-	portsRollback := []map[string]interface{}{}
-	var reserveLog *bytes.Buffer
-L:
-	for _, rr := range resourceRequests {
-		p, ok := h.pools[rr.GetResourceType()]
-		if !ok {
-			logrus.Warnf("Pool %v for host %v not found for reserving %v. Skipping reservation", rr.GetResourceType(), hostID, rr)
-			continue
-		}
-		PoolType := p.GetPoolType()
-		switch PoolType {
-		case computePool:
-			pool := p.(*ComputeResourcePool)
-			request := rr.(AmountBasedResourceRequest)
-			if !force && pool.Used+request.Amount > pool.Total {
-				err = OverReserveError{hostID: hostID, resourceRequest: rr}
-				break L
-			}
 
-			pool.Used = pool.Used + request.Amount
-			i++
-			if reserveLog == nil {
-				reserveLog = bytes.NewBufferString(fmt.Sprintf("New pool amount on host %v:", hostID))
-			}
-			reserveLog.WriteString(fmt.Sprintf(" %v total: %v used: %v.", request.Resource, pool.Total, pool.Used))
-		case portPool:
-			pool := p.(*PortResourcePool)
-			request := rr.(PortBindingResourceRequest)
-			result, e := PortReserve(pool, request)
-			portsRollback = append(portsRollback, result)
-			logrus.Debugf("Host-UUID %v, PortPool Map tcp %v, PortPool Map udp %v, Ghost Map tcp %v, Ghost Map udp %v", hostID, pool.PortBindingMapTCP, pool.PortBindingMapUDP, pool.GhostMapTCP, pool.GhostMapUDP)
-			if e != nil && !force {
-				err = e
-				break L
-			} else {
-				if _, ok := data[request.Resource]; !ok {
-					data[request.Resource] = []map[string]interface{}{}
-				}
-				data[request.Resource] = append(data[request.Resource].([]map[string]interface{}), result)
-			}
-		}
-	}
+	executedActions := []ReserveAction{}
 
-	if err == nil {
-		if reserveLog != nil {
-			logrus.Info(reserveLog.String())
-		}
-	} else {
-		logrus.Error(err)
-		// rollback
-		for _, rr := range resourceRequests[:i] {
-			p, ok := h.pools[rr.GetResourceType()]
-			if !ok {
-				break
-			}
-			resourcePoolType := p.GetPoolType()
-			switch resourcePoolType {
-			case computePool:
-				pool := p.(*ComputeResourcePool)
-				request := rr.(AmountBasedResourceRequest)
-				pool.Used = pool.Used - request.Amount
-			}
-		}
-		// roll back ports
-		if pool, ok := h.pools["portReservation"].(*PortResourcePool); ok {
-			for _, prb := range portsRollback {
-				if portReservation, ok := prb[allocatedIPs].([]map[string]interface{}); ok {
-					for _, portReserved := range portReservation {
-						ip := portReserved[allocatedIP].(string)
-						port := portReserved[publicPort].(int64)
-						prot := portReserved[protocol].(string)
-						pool.ReleasePort(ip, port, prot, "")
-						logrus.Infof("Roll back ip [%v] and port [%v]", ip, port)
-					}
-				}
+	for _, action := range reserveActions {
+		err := action.Reserve(s, resourceRequests, nil, h, force, data)
+		if err != nil {
+			logrus.Error("Error happens in reserving resource. Rolling back the reservation")
+			action.RollBack(s, resourceRequests, nil, h)
```

Comment:
Small thing: Since you always rollback the action that just errored out and then rollback the previously completed ones, could you just move `executedActions = append(executedActions, action)` to above if err != nil and then eliminate the `action.RollBack(s, resourceRequests, nil, h)` line and just go right into the rollback for loop.

Make sense?